api/mapping/views.py

```diff
@@ -545,35 +646,171 @@ def password_reset_request(request):
+
+        # Load in uploaded data dictionary for joining (From the Documents section of the webapp)
+        external_dictionary = pd.read_csv(os.path.join('media/', files[0]))
+
+        # # Create an intermediate join table
+        # # This ensures that each field in scan_report has a field description from the external dictionary
+        field_join = pd.merge(dict_df, external_dictionary, how='left', left_on='Field', right_on='Column Name')
+        field_join_grp = field_join.groupby(['Field', 'Value']).first().reset_index()
+
+        field_join_grp = field_join_grp[['Table', 'Field', 'Value', 'Frequency', 'FieldDesc', 'Column Description']]
+
+        field_join_grp.to_csv('/data/TEMP_field_join_output.csv')
+
+        # Join the intermediate join back to the external dictionary
+        # This time on field and value
+        x = pd.merge(field_join_grp, external_dictionary, how='left', left_on=['Field', 'Value'], right_on=['Column Name', 'ValueCode'])
+        x = x[['Table', 'Field', 'Value', 'Frequency', 'FieldDesc', 'Table Name', 'Column Name', 'Column Description_x', 'ValueCode', 'ValueDescription']]
+
+        x.columns = [
+            "Source_Table",
+            "Source_Field",
+            "Source_Value",
+            "Source_Frequency",
+            "Source_FieldDesc",
+            "Dictionary_TableName",
+            "Dictionary_ColumnName",
+            "Dictionary_ColumnDesc",
+            "Dictionary_ValueCode",
+            "Dictionary_ValueDescription"
+        ]
+
+        # If data are missing from imported dictionary
+        # replace with analagous descriptions to flesh out dictionary for Usagi
+        bad_index = x["Dictionary_ValueDescription"].isnull()
+        x["Dictionary_ValueDescription"][bad_index] = x["Dictionary_ValueCode"][bad_index]
+
+        bad_index = x["Source_FieldDesc"].isnull()
+        x["Source_FieldDesc"][bad_index] = x["Source_Field"][bad_index]
+
+        x.to_csv('/data/TEMP_full_join.csv')
+
+        for index, row in x.iterrows():
+            
+            print(row['Source_Field'])
+            print(row['Source_Value'])
+
+            obj = DataDictionary.objects.get(
+                source_value__scan_report_field__name=row['Source_Field'],
+                source_value__value=row['Source_Value']
+                )
+
+            print(type(obj))
+
+            # If user has fixed the definition in the webapp
+            # don't overwrite the held definition with the external dictionary values
+            if obj.definition_fixed:
+                continue
+
+            else:
+                obj.dictionary_table=row['Dictionary_TableName']
+                obj.dictionary_field=row['Dictionary_ColumnName']
+                obj.dictionary_field_description=row['Dictionary_ColumnDesc']
+                obj.dictionary_value_code=row['Dictionary_ValueCode']
+                obj.dictionary_value_description=row['Dictionary_ValueDescription']
+                obj.save()
+   
```

Comment:
happy to trust you on all this :D 
maybe i have some suggestions to simplify, but dont think this is the time

Reply:
Any suggestions welcomed. Written this super verbose so I'm crystal clear about what's being saved.

Reply:
we can add it to a backlog in the far future to have a code review / cleanup/ part of a hackathon